src/storage/txn/scheduler.rs

```diff
@@ -112,177 +172,214 @@ impl Scheduler {
+                            Ok(Some(x)) => res.push(Ok((k.raw().unwrap(), x))),
+                            Ok(None) => {}
+                            Err(e) => res.push(Err(StorageError::from(e))),
                         }
-                        Ok(res)
-                    }
-                    Err(e) => Err(e.into()),
-                };
-                callback.take().unwrap()(res);
-            }
-            Command::Scan { ref start_key, limit, start_ts, ref mut callback, .. } => {
-                let snap_store = SnapshotStore::new(snapshot, start_ts);
-                let mut scanner = try!(snap_store.scanner());
-                let key = start_key.clone();
-                let res = match scanner.scan(key, limit) {
-                    Ok(mut results) => {
-                        Ok(results.drain(..).map(|x| x.map_err(::storage::Error::from)).collect())
                     }
-                    Err(e) => Err(e.into()),
-                };
-                callback.take().unwrap()(res);
+                    ProcessResult::MultiKvpairs { pairs: res }
+                }
+                Err(e) => ProcessResult::Failed { err: StorageError::from(e) },
             }
-            Command::Prewrite { ref ctx, ref mutations, ref primary, start_ts, .. } => {
-                let mut txn = MvccTxn::new(snapshot, start_ts);
-                let mut results = vec![];
-                for m in mutations {
-                    match txn.prewrite(m.clone(), primary) {
-                        Ok(_) => results.push(Ok(())),
-                        e @ Err(MvccError::KeyIsLocked { .. }) => {
-                            results.push(e.map_err(Error::from))
+        }
+        Command::Scan { ref start_key, limit, start_ts, .. } => {
+            let snap_store = SnapshotStore::new(snapshot.as_ref(), start_ts);
+            let scanner = snap_store.scanner();
+            match scanner {
+                Ok(mut scanner) => {
+                    let key = start_key.clone();
+                    match scanner.scan(key, limit) {
+                        Ok(mut results) => {
+                            let pairs = results.drain(..)
+                                .map(|x| x.map_err(StorageError::from))
+                                .collect();
+                            ProcessResult::MultiKvpairs { pairs: pairs }
                         }
-                        Err(e) => return Err(Error::from(e)),
+                        Err(e) => ProcessResult::Failed { err: StorageError::from(e) },
                     }
                 }
+                Err(e) => ProcessResult::Failed { err: StorageError::from(e) },
+            }
+        }
+        _ => panic!("unsupported read command"),
+    };
 
-                let pr: ProcessResult = ProcessResult::ResultSet { result: results };
-                let cb = make_write_cb(pr, cid, self.schedch.clone());
+    if let Err(e) = ch.send(Msg::ReadFinished { cid: cid, pr: pr }) {
```

Comment:
If this happen, some data may stay in memory for ever. Should add a todo to clean it up.